src/directLineStreaming.ts

```diff
@@ -0,0 +1,345 @@
+      let request = BFSE.StreamingRequest.create('POST', '/v3/directline/conversations/' + this.conversationId + '/activities');
+      request.setBody(JSON.stringify(activity));
+      this.streamConnection.send(request)
+        .then((resp) => {
+          if (resp.statusCode !== 200) throw new Error("PostActivity returned " + resp.statusCode);
+          let numberOfStreams = resp.streams.length;
+          if (numberOfStreams !== 1) throw new Error("Expected one stream but got " + numberOfStreams)
+          resp.streams[0].readAsString().then((idString) => {
+            let idObj = JSON.parse(idString);
+            return subscriber.next(idObj.Id)
+          });
+        })
+        .catch((e) => {
+          // If there is a network issue then its handled by
+          // the disconnectionHandler. Everything else can
+          // be retried
+          console.warn(e);
+          return subscriber.error(e);
+        });
+    });
+    return resp$;
+  }
+
+  private postMessageWithAttachments(message: Message) {
+    const { attachments, ...messageWithoutAttachments } = message;
+
+    let httpContentList = [];
+    return Observable.create(subscriber => {
+      return Observable.from(attachments || [])
+        .flatMap((media: Media) =>
+          Observable.ajax({
+            method: "GET",
+            url: media.contentUrl,
+            responseType: 'arraybuffer'
+          })
+            .do(ajaxResponse => {
+              let buffer = new Buffer(ajaxResponse.response);
+              let stream = new BFSE.SubscribableStream();
+              stream.write(buffer);
+              let httpContent = new BFSE.HttpContent({ type: media.contentType, contentLength: buffer.length }, stream);
+              httpContentList.push(httpContent);
+            }))
+        .count()
+        .flatMap(_ => {
+          let url = `/v3/directline/conversations/${this.conversationId}/users/${messageWithoutAttachments.from.id}/upload`;
+          let request = BFSE.StreamingRequest.create('PUT', url);
+          var activityStream = new BFSE.SubscribableStream();
+          activityStream.write(JSON.stringify(messageWithoutAttachments), 'utf-8');
+          request.addStream(new BFSE.HttpContent({ type: "application/vnd.microsoft.activity", contentLength: activityStream.length }, activityStream));
+          httpContentList.forEach(e => request.addStream(e));
+          return this.streamConnection.send(request);
+        })
+        .do(resp => {
+          if (resp.streams && resp.streams.length !== 1) {
+            subscriber.error("Invalid stream count " + resp.streams.length);
+          } else {
+            resp.streams[0].readAsJson()
+              .then(json => {
+                subscriber.next(json['Id'])
+              })
+          }
+        })
+        .subscribe(_ => _) // force execution
+    })
+  }
+
+  private disconnectionHandler(e: any) {
+    if (this.connectionStatus$.value === ConnectionStatus.Connecting ||
```

Comment:
Why we don't disconnect when it is connecting?

Reply:
Yup, the check seems unnecessary.